### ISSUE TITLE
Internal "The lists had different lengths" exception

### ISSUE THREAD

**Issue by jwosty** [Microsoft/visualfsharp#5468]
I was in the process of writing a piece of code when at a certain state it started crashing the F# compiler. The code is not syntactically complete or well-typed F# code, but it shouldn't cause:

```
  error FS0193 : internal error : The lists had different lengths.list2 is 1 element shorter than list1Parameter name: list2 [/Users/jwostenberg/Code/BF.Net/BF.Net/BF.Net.fsproj]
  error FS0073 : internal error : The lists had different lengths.list2 is 1 element shorter than list1Parameter name: list2 (ArgumentException) [/Users/jwostenberg/Code/BF.Net/BF.Net/BF.Net.fsproj]
  error FS0193 : internal error : The lists had different lengths.list2 is 1 element shorter than list1Parameter name: list2 [/Users/jwostenberg/Code/BF.Net/BF.Net/BF.Net.fsproj]
```

`parserfs`:

```fsharp
module BF.Net
open System
open FSharp.Core.Printf

let Errorf format = ksprintf (fun x -> (Error x)) format

type SimpleInstruction = | Right | Left | Inc | Dec | Read | Print
type Instruction = SimpleInstruction of SimpleInstruction | Loop of Body
and Body = Body of Instruction list

let validChars = ['>'; '<'; '+'; '-'; '.'; ','; '['; ']']
let lex input = input |> Seq.filter (fun chr -> List.contains chr validChars) |> Seq.toList

let parseSimpleInstruction input =
    match input with
    | '>'::rest -> Ok (Right, rest)
    | '<'::rest -> Ok (Left, rest)
    | '+'::rest -> Ok (Inc, rest)
    | '-'::rest -> Ok (Dec, rest)
    | '.'::rest -> Ok (Print, rest)
    | ','::rest -> Ok (Read, rest)
    | symbol::_ -> Errorf "Unexpected token '%c'; expecting symbol." symbol
    | [] -> Errorf "Unexpected end of input; expecting symbol."

let rec parseLoop input =
    match input with
    | [] -> Body []
    | '['::rest ->
        match parseBody rest with
        | Error _ as err -> err
        | Ok (contents, rest) ->
            match input with
            | ']'::rest -> ()

and parseBody input = Body
```

(at the moment of error I had been working on the `parseLoop` and `parseBody` functions)

Additionally the IDE (VSfM) is highlighting the entire file in red.

#### Expected behavior

Expected: type/parsing errors, but no internal errors.

#### Actual behavior

Error present.

#### Known workarounds

Finish the code I'm writing and ignore the errors.

#### Related information 

MacOS
F# compiler 4.1
FSharp.Core 4.3.3
Mono
Visual Studio for Mac

**Comment by jwosty** [comment]
I've managed to get a really small reproducible example:

```fsharp
let rec f1 input =
    match f2 rest with
    | 42 -> ()
and f2 input = Ok
```

Some variations, some of which exhibit the bug and others of which don't:

```fsharp
// BUG PRESENT
// can be any curried DU constructor
type T = T of int
let rec f1 input =
    match f2 rest with
    | 42 -> ()
and f2 input = T
```

```fsharp
// BUG NOT PRESENT
// so it has to specifically be a curried DU constructor, not just any curried function...
let ok = Ok
let rec f1 input =
    match f2 rest with
    | 42 -> ()
and f2 input = Ok
```

```fsharp
// BUG NOT PRESENT
// can't be a wildcard pattern
let rec f1 input =
    match f2 rest with
    | _ -> ()
and f2 input = Ok
```

```fsharp
// BUG NOT PRESENT
let f2 input = Ok
let rec f1 input =
    match f2 rest with
    | 42 -> ()
```

```fsharp
// BUG PRESENT
// works when you give it an annotation indicating (f2 rest) isn't a function (but it really is)
let rec f1 input =
    match f2 rest with
    | 42 -> ()
and f2 input: 'Result<_,_> = Ok
```

```fsharp
// BUG NOT PRESENT
// disappears when you annotate f2 more accurately
let rec f1 input =
    match f2 rest with
    | 42 -> ()
and f2 input: _ -> 'Result<_,_> = Ok
```

**Comment by forki** [comment]
I tried to repro with the first small sample but seems to work. Where is
rest coming from?

John Wostenberg <notifications@github.com> schrieb am Mi., 8. Aug. 2018,
23:16:

**Comment by jwosty** [comment]
@forki sorry what are you asking?

**Comment by forki** [comment]
Can you please create a code snippet that repros in fsi? I tried to repro
with the stuff you put in last comment and didn't get that internal error

John Wostenberg <notifications@github.com> schrieb am Do., 9. Aug. 2018,
16:24:

**Comment by jwosty** [comment]
@forki ah. I can't get it to happen in fsi either.

**Comment by forki** [comment]
Ok. That's weird. I guess it's still happening, but fsi is not showing it.

John Wostenberg <notifications@github.com> schrieb am Do., 9. Aug. 2018,
16:31:

**Comment by jwosty** [comment]
Yes, interesting -- I can still continue the session after the bad code so maybe it's swallowing the exception somewhere.

**Comment by forki** [comment]
Ok. Can you please upload a minimal fs file that repros for you?

John Wostenberg <notifications@github.com> schrieb am Do., 9. Aug. 2018,
16:35:

**Comment by jwosty** [comment]
Yikes... I've gotten it even smaller. Just paste this into a .fs file and try to compile it:

```fsharp
let f x : _ option = Some
```

**Comment by forki** [comment]
ok I can repro.

Am Do., 9. Aug. 2018 um 16:41 Uhr schrieb John Wostenberg <
notifications@github.com>:

**Comment by jwosty** [comment]
@forki well that's good and bad.

**Comment by forki** [comment]
@dsyme I looked at it, but I don't know how this is supposed to work. sorry

**Comment by nikonthethird** [comment]
I also came across this "The lists had different lengths" error. I don't know it's the same one, but the repro is really tiny:

```
type Test = Test of Unit
let test : Test = Test
```

![error](https://user-images.githubusercontent.com/15893812/44163759-8cbc8780-a0c4-11e8-897e-002b59e02f1f.jpg)

I updated to VS 15.8 today, so I can't test if this happened in F# 4.1 also...

**Comment by jwosty** [comment]
@nikonthethird yes, this seems to be the same bug. Also, I've been testing this on Mono + F# 4.1, so it's been there and hasn't been fixed yet.

**Comment by forki** [comment]
@kevinransom @dsyme I looked at this code, but it's really hard to tell why
the assertion is in place and why it's not using a safe error handler
instead. Also I'm sure the root cause is somewhere else and it's just that
it breaks the assumptions here.

John Wostenberg <notifications@github.com> schrieb am Mi., 15. Aug. 2018,
19:53:

**Comment by cartermp** [comment]
We investigated and know why this is happening. In both cases:

```fsharp
type Test = Test of Unit
let test : Test = Test
```
and

```fsharp
let rec parseLoop input =
    match input with
    | [] -> Body []
    | '['::rest ->
        match parseBody rest with
        | Error _ as err -> err
        | Ok (contents, rest) ->
            match input with
            | ']'::rest -> ()

and parseBody input = Body
```

The types are actually mismatched as @jwosty has mentioned. In the first case, `Test` is the declared type, but what's being written is a partially applied `Test` constructor. In the second case, `parseLoop` infers the type to be `Body of instruction list`, but `parseBody` is declared to be a partially applied DU constructor.

However, the code never makes it far enough into the compiler to make this apparent, so it just blows up with the internal assertion (and makes the whole editor red squiggles).

@TIHan is working out a proper fix to this so that we can surface the type mismatch error.

We're not yet sure, but this may go back as far as F# 4.0.

**Comment by cartermp** [comment]
Actually, it looks like some change between F# 4.0 and now has caused this assertion to trigger: https://repl.it/@cartermp/GlisteningBoilingQuagga

```
"F# Compiler for F# 4.0 (Open Source Edition)
   
exit status 1

/home/runner/main.fs(2,19): error FS0001: This expression was expected to have type
    Test    
but here has type
    Unit -> Test 
```

**Comment by forki** [comment]
No. The type error is emitted as well. Actually if you run it in fsi you only see the type error. I think this is running in some error recovering scenario.

**Comment by cartermp** [comment]
Yep, you're right - that's what's going on. Something was fiddled with in there. Another oddity:

```fsharp
let doot : int = fun () -> 1
```

This doesn't give a type mismatch error, it gives

**Comment by TIHan** [comment]
Tested in FSI, I only see the type error. This is indeed interesting.

I have a possible fix. I need to look a bit more into before I make  PR though.

**Comment by TIHan** [comment]
I know the PR that caused this: https://github.com/Microsoft/visualfsharp/pull/1568

At this line here: https://github.com/Microsoft/visualfsharp/pull/1568/files#diff-5b9ab9dd9d7133aaf23add1048742031R8100

It's basically eating the exception when type checking fails and we continue to do what we need to do, basically to fix this: https://github.com/Microsoft/visualfsharp/issues/629

We should review this PR again to make sure this is the right thing to do. But I still have a potential fix that doesn't involve messing with those changes.
